torch/testing/_internal/logging_tensor.py

```diff
@@ -74,52 +94,100 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
 class LoggingTensorReentrant(LoggingTensor):
     context = torch.overrides.enable_reentrant_dispatch
 
+def _maybe_repr(a):
+    return a if isinstance(a, str) else repr(a)
+
 # https://stackoverflow.com/questions/36408496/python-logging-handler-to-append-to-list
 class LoggingTensorHandler(logging.Handler):
-    log_list: List[str]
-    next_shortid: int
-
-    def __init__(self, log_list: List[str], use_shortid_for_all_tensors: bool) -> None:
+    def __init__(
+            self, log_list: List[str], use_shortid_for_all_tensors: bool,
+            with_type: bool, tracebacks_list: Optional[List]) -> None:
         logging.Handler.__init__(self)
         self.log_list = log_list
-        self.next_shortid = 0
         self.use_shortid_for_all_tensors = use_shortid_for_all_tensors
-
-    # WARNING: not deterministic over multiple threads, this matters for
-    # autograd
-    def _shortid(self, o: object) -> int:
-        if not hasattr(o, '_shortid'):
-            o._shortid = self.next_shortid  # type: ignore[attr-defined]
-            self.next_shortid += 1
-        return o._shortid  # type: ignore[attr-defined]
-
-    def _fmt(self, a: object) -> str:
+        self.tracebacks_list = tracebacks_list
+        self.memo = {}
+        self.next_id = 0
+        self.with_type = with_type
+
+    def _shortid(self, t: torch.Tensor) -> int:
+        o = WeakIdRef(t)
+        weak_self = weakref.ref(self)
+
+        def del_memo():
+            self = weak_self()
+            if self is None:
+                return
+            self.memo.pop(o, None)
+
+        weakref.finalize(t, del_memo)
```

Comment:
Why do this by hand and no use `torch.utils.weak.WeakTensorKeyDictionary` for self.memo?